### PR TITLE
Reverts around time_zone

### PR DESCRIPTION
**Pull request by fmayer**
This broke multiple buildbots.
See the comments in https://github.com/llvm/llvm-project/pull/89537.

I intend to submit those as separate changes, not as one squashed change.

### CODE REVIEW

**Comment by ldionne** [comment]
Can you please provide context for why you want to revert these patches?

**Comment by fmayer** [comment]
> Can you please provide context for why you want to revert these patches?

See the comments in https://github.com/llvm/llvm-project/pull/89537. This broke multiple buildbots.

**Review by vitalybuka — APPROVED** [review]

**Comment by fmayer** [comment]
> "lowest effort revert back to green without enough information to then move forward" kind of thing.

This is not only our bot, but others as well.

Also, quoting LLVM policy:

> If you break a buildbot in a way which can’t be quickly fixed, please revert.

It's not about effort, it's about not stacking breakages on top of each other. It's infeasible to add all these configurations to pre-commit UI. Some of our bots take hours to compile multi-stage LLVM with sanitizers.

**Comment by vitalybuka** [comment]
> I'd like @mordante to chime in before these patches are reverted. It's not clear to me that we currently have a path to fixing this after the revert, so this feels a lot like a "lowest effort revert back to green without enough information to then move forward" kind of thing.

It does not work with different timezones. And seems @mordante  didn't reply.

Revert and re-land is very trivial. Why do we need to have expection for libc++ when this approach work just fine for entire LLVM.

> By the way, these configurations are also not visible in our pre-commit CI. Please add them there, that is our policy to get libc++ support.

Do you have arm hardware?

**Comment by mordante** [comment]
> > "lowest effort revert back to green without enough information to then move forward" kind of thing.
> 
> This is not only our bot, but others as well.

The HWSAN is fixed by disabling the test in the CI right? I got an e-mail of the sanitizer failure, however in my experience the sanitizer builds tend to be flaky so I typically look whether they fail for multiple builds. (I spend a lot of time in the past to investigate build failures and the next build was green again.)

I have not gotten e-mails of other build failures.

> Also, quoting LLVM policy:
> 
> > If you break a buildbot in a way which can’t be quickly fixed, please revert.
> 
> It's not about effort, it's about not stacking breakages on top of each other. 

Louis asked how we can then move forward. That is part of the same policy

"It is not considered reasonable to revert without at least the promise to provide a means for the patch author to debug the root issue. If a situation arises where a public reproducer can not be shared for some reason (e.g. requires hardware patch author doesn’t have access to, sharp regression in compile time of internal workload, etc.), the reverter is expected to be proactive about working with the patch author to debug and test candidate patches."

So I would like to know how we can move ahead before reverting. Since these changes can depend on the system configuration it's hard to debug them on these systems.

**Comment by fmayer** [comment]
> I got an e-mail of the sanitizer failure, however in my experience the sanitizer builds tend to be flaky

It's been continuously failing since yesterday.

> Louis asked how we can then move forward. That is part of the same policy

You can run the sanitizer buildbot configuration, as long as you have arm64 hardware. It says in the buildbot error message:

> How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild

**Comment by vitalybuka** [comment]
>> experience the sanitizer builds tend to be flaky 

Nothing in sanitizers that can cause flakiness on it's own.
Usually it happens for long tests, so we don't like long tests.

> > How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild

Technically it should be enough to reproduce.
We also can promise to debug this on our bot.

But if it's just a slow test, as we have an evidence for that, can you just split it some how soon, to cut slowest parts e.g. in half?

> The HWSAN is fixed by disabling the test in the CI right?

Somehow it's not disabled. Probably something in libc++ test infra didn't propagate hwasan feature.

**Comment by vitalybuka** [comment]
> > The HWSAN is fixed by disabling the test in the CI right?
> 
> Somehow it's not disabled. Probably something in libc++ test infra didn't propagate hwasan feature.

@fmayer Just niticed, XFAIL does not fix timouts, as XFAIL executes test anyway. It must be UNSUPPORTED

**Comment by ldionne** [comment]
It's easy to quote stuff from the Developer Policy out of context. The Developer Policy basically states that it's OK to revert to a green state to keep the CI afloat, and it describes provisions for how people can work together to make progress.

It doesn't condone reverting a project's work without a reasonable path to fixing the bots, nor does it state that arbitrary people can set up arbitrary build bot configurations (on arbitrary hardware and arbitrary machines) and then hold project contributors accountable for failures on these bots to the point of not being able to make progress because of them.

Libc++ has a documented support policy and these bots seem to be falling outside of that. I'm not going to play games with you and pretend that we don't have to fix the issues on your bots: we definitely want to get that fixed ASAP. However, since these setups fall outside of our CI, are inaccessible to us and are clearly a bit unusual, we need your help to investigate the issue or provide easy access to the machine so we can reasonably investigate it ourselves. Without that, we'll basically revert these patches, folks will disappear and we'll be left without the ability to re-land the patches.

Once this is over: We work really hard to make our CI as good as possible and covers all of our supported usage scenarios. Please work with us to include your usage scenarios into our pre-commit CI, these events are a huge pain for everyone involved and can be solved with technical means.

**Comment by ldionne** [comment]
Note that if you really just want to turn off this whole feature on your configuration while we're figuring out a path forward, I think you could just pass `-DLIBCXX_ENABLE_TIME_ZONE_DATABASE=OFF` in the CMake and this should disable all these tests. @mordante should be able to confirm.

This would have the added benefit that the issue can be reproduced on your setup by just re-enabling the configuration.

**Comment by fmayer** [comment]
> to the point of not being able to make progress because of them.

It's not that at all. It's about reverting, figuring out what to do, and re-landing. Which is the norm in (at least non-libcxx) LLVM.

**Comment by mordante** [comment]
> > Louis asked how we can then move forward. That is part of the same policy
> 
> You can run the sanitizer buildbot configuration, as long as you have arm64 hardware. It says in the buildbot error message:
> 
> > How to reproduce locally: https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild

I don't have access to ARM64 hardware. I also noticed this page states "You need system similar to the buildbot, details can be found in worker info on the build page, e.g. [sanitizer-x86_64-linux-fast](https://lab.llvm.org/buildbot/#/builders/sanitizer-x86_64-linux-fast)" at the linked site I don't see the requirements how to build this. I would be helpful if that page can be improved.

Since this error is a time-out it might not be possible to reproduce on a different machine. As @vitalybuka  mentioned the test should be `UNSUPPORTED` instead of `XFAIL`.

> Note that if you really just want to turn off this whole feature on your configuration while we're figuring out a path forward, I think you could just pass `-DLIBCXX_ENABLE_TIME_ZONE_DATABASE=OFF` in the CMake and this should disable all these tests. @mordante should be able to confirm.

Yes that is the option to disable this.

**Comment by fmayer** [comment]
. > Since this error is a time-out it might not be possible to reproduce on a different machine. As @vitalybuka mentioned the test should be `UNSUPPORTED` instead of `XFAIL`.

I already submitted this.

> > Note that if you really just want to turn off this whole feature on your configuration while we're figuring out a path forward, I think you could just pass `-DLIBCXX_ENABLE_TIME_ZONE_DATABASE=OFF` in the CMake and this should disable all these tests. @mordante should be able to confirm.
> 
> Yes that is the option to disable this.

**Comment by fmayer** [comment]
> Since this error is a time-out it might not be possible to reproduce on a different machine. As @vitalybuka mentioned the test should be `UNSUPPORTED` instead of `XFAIL`.
> 

I already submitted this. I will note that even on my workstation without sanitizers, this test is slow. This revert wasn't mainly for sanitizers in the first place, but because other people also reported issues on the PR so I proposed this revert while we figure out the various issues.

**Comment by vitalybuka** [comment]
> It's easy to quote stuff from the Developer Policy out of context. The Developer Policy basically states that it's OK to revert to a green state to keep the CI afloat, and it describes provisions for how people can work together to make progress.
> 
> It doesn't condone reverting a project's work without a reasonable path to fixing the bots, nor does it state that arbitrary

Isn't should be the opposite, if way forward is not clear - revert, then figured, including working with bots maintainers. I always help if to debug on my bots, I can't promise to do that immediately.  So revert and debug AFTER 

We do this all the time on the rest of LLVM, but somehow it's a problem for libcxx. 

>  people can set up arbitrary build bot configurations (on arbitrary hardware and arbitrary machines) and then hold project contributors accountable for failures on these bots to the point of not being able to make progress because of them.
> 

I am rely grateful to people who to setup and maintain those bots. I would rather be slowed down just after landing the patch, then to look into that in 6 month when it reached release on that specific config, and now I am out of context and need to switch from something else. 

I think those arbitrarily bots very helpful in long run.

> Libc++ has a documented support policy and these bots seem to be falling outside of that. I'm not going to play games with you and pretend that we don't have to fix the issues on your bots: we definitely want to get that fixed ASAP. However, since these setups fall outside of our CI, are inaccessible to us and are clearly a bit unusual, we need your help to investigate the issue or provide easy access to the machine so we can reasonably investigate it ourselves. Without that, we'll basically revert these patches, folks will disappear and we'll be left without the ability to re-land the patches.
> 

Yes, that default, sorry if it's not clear.
After revert you should expect to get more information. Isn't how it works for LLVM in general?
E.g. last month I reverted sized new/delete in libcxx, then debugged, and proposed a patch. But debugging was not possible immediately, and we need to recover those bots.

> Once this is over: We work really hard to make our CI as good as possible and covers all of our supported usage scenarios. Please work with us to include your usage scenarios into our pre-commit CI, these events are a huge pain for everyone involved and can be solved with technical means.

Happy to help with CI, but missing CI should not be excuse for no-revert.

**Comment by vitalybuka** [comment]
I guess in general we arguing about these two approaches:

1. @fmayer reverts
2. @mordante investigates
3. @mordante asks @fmayer for help
4. @mordante relands with the fix

and 

1. @fmayer reports
2. built-bot spam all users on blame list of  unrelated patches, and spent time investigating
3. maybe we miss another regression, because bot was red anyway
4. downstream user pickup the issues and investigate
5. @mordante investigates
6. @mordante asks @fmayer for help
7. @mordante lands a fix

To me @mordante and @fmayer time investment of Revert <= No-Revert, but No1 is less likely annoy the community and downstream users.

So I strongly advocate for No1 and revert when in doubt.

**Comment by ldionne** [comment]
I think that's a false dilemma.

My understanding is that there are three failures reported:
1. Fuchsia: https://ci.chromium.org/ui/p/fuchsia/builders/toolchain.ci/clang-linux-x64/b8745592240286102897/overview (assertion failure)
2. Windows/Linux cross builders: https://lab.llvm.org/buildbot/#/builders/60/builds/17468 (assertion failure)
3. HWasan: https://lab.llvm.org/buildbot/#/builders/236/builds/11695 (timeout)

The first two issues don't have a path to being resolved because the reporters have not provided a ways to reproduce the failures or investigated them. Therefore, they are not grounds for asking a revert of the patch. This is especially true given that TZDB is a library that is inherently coupled with the system in use, which means that system differences and quirks are intended. This means that the only viable option might very well be to `XFAIL` the test on the platforms that are failing because they e.g. have an incomplete or malformed time zone database. Of course we can't know until someone has investigated on these machines.

The last issue is a timeout. The fact that you run a build bot with instrumentation that greatly slows down runtime tests and can cause them to timeout is also not grounds to revert the patch. The right solution here is to mark the test as unsupported under that configuration, or to use a faster machine, or to make the tests faster, or to improve hwasan to have less runtime impact. It doesn't make the code that was committed wrong as in "incorrect". To further explain my point: if I enable a CI machine that is vastly underpowered and causes a bunch of tests to start timing out, whose fault is it? Can I then go to patch authors and ask them to revert their patches under the Dev Policy because of my bot? I think we'll all agree that it doesn't make sense. Instead, *I* should figure out a way to make my bot robust to longer-running tests and take it upon myself to deal with the implications of running such a bot.

So, to summarize: When we ship a bug and we can't quickly fix it forward, we revert. We all agree on that, it's the LLVM way. However, many things don't qualify as "a bug" yet still break CI. This is a prime example of such a situation, and I am really exhausted of folks trying to strong-arm others into reverting patches blindly just to get their bots green again with minimal effort while citing the LLVM policy.

@vitalybuka Is there even still a problem? You marked the test as unsupported under hwasan, so it's not running under your configuration. Did that work? A better fix would be to use `--param long_tests=False` to disable all the slow-running tests in the suite instead, and to mark the `local_time` test as `REQUIRES: long_tests`.

**Comment by fmayer** [comment]
> really exhausted of folks trying to strong-arm others into reverting patches blindly

no one strong-armed anyone. From out point of view: this breaks our bot and at least two others, we don't have enough context into what you are doing to fix forward, or to know whether this is indicative a bug, so we proposed a revert. That's it. I don't know what the big fuss is about, submitting and reverting the revert would have taken less time than this meta-discussion. It will always be the case that buildbot maintainers do not have the same level of context into the change, so they will err more on the side of reverts than the author would.

**Comment by ldionne** [comment]
> no one strong-armed anyone

I consider selectively quoting the part of the LLVM Developer Policy that says "reverts are fine" to be a form of that.

This fuss is about the dynamics in LLVM that make some people think that reverting a month of work over a test that's running too slowly under a sanitizer and should be addressed with a Lit annotation. It happens a lot in libc++ due to the nature of the project running under many configurations and on many platforms, and it's exhausting. It also doesn't help that it comes almost exclusively from corporate actors and is imposed upon contributors who (in libc++) are mostly enthusiasts, leading to a bad contribution experience.

**Comment by fmayer** [comment]
Again, the revert was because it broke three buildbots, not only for the lit annotation. We don't want to mask problems, and if multiple bots fail, maybe something is actually wrong. If you say the other two buildbots are running unsupported configurations, that's okay with me, but there is no way for me to know this, it wasn't called out in the PR. For the record I will say that I too have worked on runtime components, which are by definition hard to test in all configurations, and also got my changes reverted sometimes. Please don't assume bad intend based on whether someone is a "corporate actor".

**Comment by ldionne** [comment]
> Please don't assume bad intent based on whether someone is a "corporate actor".

I did not assume bad intent based on that. BTW I am employed full-time working on LLVM myself. What I meant is simply that what can appear as acceptable to someone working on LLVM as their full-time job in a corporate context may be perceived completely differently by someone who is doing this as an enthusiast outside of a corporate setting, and that can lead to contributor alienation for those people. Libc++ has a much higher concentration of enthusiasts than any other area of LLVM (at least in my experience).

**Comment by vitalybuka** [comment]
@ldionne  I am not sure  why you are on such defense on reverts. 

>  It doesn't make the code that was committed wrong as in "incorrect".
> if I enable a CI machine that is vastly underpowered and causes a bunch of tests to start timing out, whose fault is it? 

It's not about blaming who is this fault. It's about doing necessary to maintain the code health.

It's not ONLY about correctness. Our build bots are way to keep compiler from regressing. And time to time revert of correct code is needed to keep bots functioning, maybe just for upgrading or re-configuring build steps. Keeping attention to bots is the price to pay to have good diverse test coverage and detect REAL bugs early.

For problematic bots:
1. We can decide that those bots are not helpful, and remove them from master, or move to staging, where it does not spam commiters.  
2. Or maybe that bot is extremely important we will need to figure-out something else, including delaying  landing some patches.

But it's not the case we have here. That bot is consistently green and well maintained.

> So, to summarize: When we ship a bug and we can't quickly fix it forward, we revert. We all agree on that, it's the LLVM way. However, many things don't qualify as "a bug" yet still break CI. 

And until proven there is always safe to assume that there is a bug than CI failure. In particular case possibility of 'deadlock' what is concerned me.

And my point that threshold for revert should be significantly lower than TRUE bug. Some times it's need to be done to figure out how to proceed further. If it's CI issue, it's better to revert, bot maintainers may need some time catching up, e.g. when we upgrade some deps requirement, we may revert back and forth that patch, until we upgrade all bots. It's way better then we broke all bots and fix forward them next few weeks keeping then non-functional.

Reverting is a trivial thing, it's not some terrible thing that must never happens.
In healthy work flow it should be possible to speculatively revert a patch even if there is a doubt that this is the cause.
It's always trivial to reland the patch if the revert is not helpful.

E.g. here @mordante was able quickly guess problems, and if we reverted that yesterday, we'd already had it relanded with UNSUPPORTED spending less time than we spent arguing here.

It works for the rest of LLVM. works for large OSS projects, e.g. Chromium, why libcxx needs to be special?

To summarize:
1. cost of false revert is not zero, but extremely low
1. cost of false non-revert significant 
So reverting, even when not, or even just for fun ( don't quote me on this one :) ) , should be acceptable practice. We are not in the state when someone is abusing reverts, and if we have abuser, it would be easy to resolve.

>@vitalybuka Is there even still a problem? You marked the test as unsupported under hwasan, so it's not running under your configuration. Did that work? 

Yes, it works.
I guess will well have to do the same for asan, it's inconsistently timeouts there as well https://lab.llvm.org/buildbot/#/builders/239/builds/7415
Ideally the test needs  to be speedup somehow.

> A better fix would be to use --param long_tests=False to disable all the slow-running tests in the suite instead, and to mark the local_time test as REQUIRES: long_tests.

Thanks! I didn't know. We will use this. Then disabling asan is unnecessary,

This particular issue is resolved.

My main concern is that reverts in libcxx are often an issue and more complicated than in the rest of LLVM, making bot maintenance difficult. Can we do better?

**Comment by mordante** [comment]
> @ldionne I am not sure why you are on such defense on reverts.
> 
> > It doesn't make the code that was committed wrong as in "incorrect".
> > if I enable a CI machine that is vastly underpowered and causes a bunch of tests to start timing out, whose fault is it?
> 
> It's not about blaming who is this fault. It's about doing necessary to maintain the code health.
> 
> It's not ONLY about correctness. Our build bots are way to keep compiler from regressing. And time to time revert of correct code is needed to keep bots functioning, maybe just for upgrading or re-configuring build steps. Keeping attention to bots is the price to pay to have good diverse test coverage and detect REAL bugs early.
> 
> For problematic bots:
> 
>     1. We can decide that those bots are not helpful, and remove them from master, or move to staging, where it does not spam commiters.
> 
>     2. Or maybe that bot is extremely important we will need to figure-out something else, including delaying  landing some patches.
> 
> 
> But it's not the case we have here. That bot is consistently green and well maintained.

As mentioned before, the sanitizer bots have false positives every now and then. I agree with @ldionne that a time-out in a bot is not per-se an error in our code.

> > So, to summarize: When we ship a bug and we can't quickly fix it forward, we revert. We all agree on that, it's the LLVM way. However, many things don't qualify as "a bug" yet still break CI.
> 
> And until proven there is always safe to assume that there is a bug than CI failure. In particular case possibility of 'deadlock' what is concerned me.
> 
> And my point that threshold for revert should be significantly lower than TRUE bug. Some times it's need to be done to figure out how to proceed further. If it's CI issue, it's better to revert, bot maintainers may need some time catching up, e.g. when we upgrade some deps requirement, we may revert back and forth that patch, until we upgrade all bots. It's way better then we broke all bots and fix forward them next few weeks keeping then non-functional.
> 
> Reverting is a trivial thing, it's not some terrible thing that must never happens. In healthy work flow it should be possible to speculatively revert a patch even if there is a doubt that this is the cause. It's always trivial to reland the patch if the revert is not helpful.

Reverting and relanding is not always trivial. Temporary disabling tests on configurations that do not work is a solution we use in libc++.
 
> E.g. here @mordante was able quickly guess problems, and if we reverted that yesterday, we'd already had it relanded with UNSUPPORTED spending less time than we spent arguing here.
> 
> It works for the rest of LLVM. works for large OSS projects, e.g. Chromium, why libcxx needs to be special?
> 
> To summarize:
> 
>     1. cost of false revert is not zero, but extremely low
> 
>     2. cost of false non-revert significant
>        So reverting, even when not, or even just for fun ( don't quote me on this one :) ) , should be acceptable practice. We are not in the state when someone is abusing reverts, and if we have abuser, it would be easy to resolve.
> 
> 
> > @vitalybuka Is there even still a problem? You marked the test as unsupported under hwasan, so it's not running under your configuration. Did that work?
> 
> Yes, it works. I guess will well have to do the same for asan, it's inconsistently timeouts there as well https://lab.llvm.org/buildbot/#/builders/239/builds/7415 Ideally the test needs to be speedup somehow.
> 
> > A better fix would be to use --param long_tests=False to disable all the slow-running tests in the suite instead, and to mark the local_time test as REQUIRES: long_tests.
> 
> Thanks! I didn't know. We will use this. Then disabling asan is unnecessary,
> 
> This particular issue is resolved.
> 
> My main concern is that reverts in libcxx are often an issue and more complicated than in the rest of LLVM, making bot maintenance difficult. Can we do better?

We try to do better, in libc++ we have a pre-commit CI to test all our supported configurations. This means patches in libc++ typically have been tested in the CI more than typical patches in the LLVM project. My personal issue with reverting directly is that it sometimes is easier to add a patch that temporary disables a configuration. Especially since buildbot maintainers sometimes don't have time to look at an issue in a timely fashion. That means development of a feature is blocked until that can be resolved. (Using stacked patches in GitHub, unlike in Phabricator, is quite painful. This is already makes development of larger features harder than previously.)

Note that in cases where I expect issues in the code due to CI issues I revert my patches. In this case I only got 1 e-mail of a failing build, not of all 3 failing builds.

**Comment by vitalybuka** [comment]
> As mentioned before, the sanitizer bots have false positives every now and then. I agree with @ldionne that a time-out in a bot is not per-se an error in our code.

In this case it was ~10 consecutive failures, which is definitely not a flake.

I will be very concerned about health of the project if we start treating bot issues as not "out code".

> Reverting and relanding is not always trivial. Temporary disabling tests on configurations that do not work is a solution we use in libc++.

Yes. But most of the time it's trivial.
Temporarily disabling is absolutely fine. But that's not the point. We should not think twice before reverting.

> We try to do better, in libc++ we have a pre-commit CI to test all our supported configurations.

CI does best effort but can't cover all supported configurations. E.g. I don't see anything except x86.

> My personal issue with reverting directly is that it sometimes is easier to add a patch that temporary disables a configuration.

That all true, but reverting should still be treated as trivial thing.
Sorry, here I made wrong call about bug or not, suspecting deadlock, but it's not the point. If we revert, after day or so of  debugging, or your insight about long test, we'd reland and move on. Note this was landed in about a week after approval, I don't buy that 1-2 days will delay will make a difference. 

> Especially since buildbot maintainers sometimes don't have time to look at an issue in a timely fashion. That means development of a feature is blocked until that can be resolved.

"Blocked" is exaggeration. We all work we a lot of branches and commits in local checkouts.

>  (Using stacked patches in GitHub, unlike in Phabricator, is quite painful. This is already makes development of larger features harder than previously.)

https://github.com/getcord/spr not the same but pretty close.

BTW. It would be nice to at least pause landing the stack after the firs bot email.

> Note that in cases where I expect issues in the code due to CI issues I revert my patches. 

If you believed that this is CI issue from the beginning, it could be helpful to comment. Somehow it didn't click for me from the beginning that this is just a long test, and not deadlock, which means a bug in the compiler or libcxx.

But again it's not the point. If we do reverts easily, we had this patch re-landed with test disabled in no time, without the drama.
